File: cyacas/libyacas/src/xmltokenizer.cpp
#include "yacas/xmltokenizer.h"
#include "yacas/lisperror.h"

#include <cctype>
#include <iostream>

std::string XmlTokenizer::NextToken(LispInput& aInput)
{
    if (aInput.EndOfStream())
        return "";

    std::string leading_spaces;
    while (std::isspace(aInput.Peek()))
        leading_spaces.push_back(aInput.Next());

    if (aInput.EndOfStream())
        return "";

    std::string s;

    char c = aInput.Next();
    s.push_back(c);

    if (c == '<') {
        while (c != '>') {
            if (aInput.EndOfStream())
                throw LispErrCommentToEndOfFile();

            c = aInput.Next();

            s.push_back(c);
        }
    } else {
        while (aInput.Peek() != '<' && !aInput.EndOfStream())
            s.push_back(aInput.Next());
        s = leading_spaces + s;
    }

    return s;
}

File: jyacas/net/sf/yacas/XmlTokenizer.java
package net.sf.yacas;

class XmlTokenizer extends LispTokenizer {
    @Override
    public String NextToken(LispInput aInput, LispHashTable aHashTable) throws Exception {

        char c;

        if (aInput.EndOfStream())
            return aHashTable.LookUp("");

        StringBuilder leadingSpaces = new StringBuilder();
        while (Character.isWhitespace(aInput.Peek()))
            leadingSpaces.append(aInput.Next());

        if (aInput.EndOfStream())
            return aHashTable.LookUp("");

        StringBuilder token = new StringBuilder();

        c = aInput.Next();
        token.append(c);

        if (c == '<') {
            while (c != '>') {
                LispError.Check(!aInput.EndOfStream(), LispError.KLispErrCommentToEndOfFile);
                c = aInput.Next();
                token.append(c);
            }
        } else {
            while (aInput.Peek() != '<' && !aInput.EndOfStream()) {
                c = aInput.Next();
                token.append(c);
            }
            leadingSpaces.append(token);
            token = leadingSpaces;
        }
        return aHashTable.LookUp(token.toString());
    }
}
